### ISSUE TITLE
Feature: Extend viewport beyond bounds to show entire bounded area (allow 'underzooming')

### ISSUE THREAD

**Issue by larsmaxfield** [maplibre/maplibre-gl-js#4510]
## Original issue

If the viewport is not square and `renderWorldCopies = false`, the map does not let you zoom out enough to show the entire map.

You can see this in the [render world copies example](https://maplibre.org/maplibre-gl-js/docs/examples/render-world-copies/) by fully zooming out and toggling the option:

![with-copies](https://github.com/user-attachments/assets/817bb546-5c60-4097-bb08-b3f7c4577031)

![no-copies-wide](https://github.com/user-attachments/assets/f83ee4ac-936c-48a1-bcbf-6d1883db488f)

<img width="223" height="393" alt="no-copies-tall" src="https://github.com/user-attachments/assets/64775ba0-8967-4091-8ee9-bbbb34b76b6d" />

## Problem

The problem is that MapLibre confines the viewport movement to within the map's bounds. The viewport cannot zoom out or pan to show areas outside the map's bounds.

When the aspect ratio of the viewport is different than that of the bounds, the viewport is effectively shrunk to fit the bounds.

Therefore, in the case of the un-bounded single-copy square map, a _non-square_ viewport cannot be zoomed out to show the entire map. Otherwise the viewport would reveal areas outside the bounds.

## Proposal

I would like an option to extend the viewport beyond the bounds such that you can see the entire bounded area regardless of aspect ratio. This would allow you to 'underzoom' and 'overpan' the map (to use the terminology I wrote in the original unedited issue).

The responses in this thread support such a feature. See also the (duplicate) request #4591.

This would only apply when `renderWorldCopies = false`, whereby the background area outside the map is transparent:

![desired](https://github.com/user-attachments/assets/a064a3d0-ebbb-4bfd-97fa-12bca114b4bf)

## API changes

My latest proposal is to add options to the map like this:

```js
const map = new maplibregl.Map({
    renderWorldCopies: false,
    ...,
    viewportBoundsContraint: 'extend',  // Default 'confine'
    extendScaleLimit: 50,  // You can zoom out until the bounds are 50% the viewport size
    extendPanLimit: 25  // You can pan until the bounds are 25% the distance to viewport center
    });
```

(One could think of this as (un)constraining the zoom and pan of the map, which the OpenSeadragon image viewer does nicely here: https://openseadragon.github.io/examples/ui-zoom-and-pan/.)

**Comment by larsmaxfield** [comment]
The currently restricted zoom-out behavior is _not_ intended if I understand [the renderWorldCopies docs](https://maplibre.org/maplibre-gl-js/docs/API/type-aliases/MapOptions/#renderworldcopies) correctly:

> When the map is zoomed out far enough that a single representation of the world does not fill the map's entire container, there will be blank space beyond 180 and -180 degrees longitude.

**Comment by sbachinin** [comment]
Restricted behaviour is intended, despite this quote.
Making it non-restricted looked a bit problematic. Also I assumed that viewing the whole globe is an unlikely demand.
Could you please describe the use case where it is necessary?

**Comment by larsmaxfield** [comment]
I was able to implement this by modifying a few lines of `getConstrained()` in `transform.ts`. Works nice on a 2D map:

![under-zooming-2d](https://github.com/user-attachments/assets/01cfca84-0e04-4e2b-97ce-7b5bb125e724)

However, on a 3D terrain map, it locks up when under-zoomed below 0 or when pitching at zoom levels below 2, sometimes throwing an 'outside of bounds' error:
```
tile_id.ts:21  Uncaught Error: x=0, y=2, z=0.27 outside of bounds. 0<=x<1.2058078276907604, 0<=y<1.2058078276907604 0<=z<=25 
    at CanonicalTileID (tile_id.ts:21:19)
    at new OverscaledTileID (tile_id.ts:96:26)
    at Terrain._getOverscaledTileIDFromLngLatZoom (terrain.ts:455:24)
    at Terrain.getElevationForLngLatZoom (terrain.ts:187:53)
    at Map._elevateCameraIfInsideTerrain (camera.ts:1122:42)
    at camera.ts:1144:39
    at Map._applyUpdatedTransform (camera.ts:1161:17)
    at HandlerManager._fireEvents (handler_manager.ts:594:23)
    at HandlerManager.stop (handler_manager.ts:309:14)
    at HandlerManager.handleEvent (handler_manager.ts:359:18)
```

**Comment by larsmaxfield** [comment]
> Restricted behaviour is intended, despite this quote. Making it non-restricted looked a bit problematic. Also I assumed that viewing the whole globe is an unlikely demand. Could you please describe the use case where it is necessary?

The general use case is for when users wish to see the entire map no matter the aspect ratio of the viewport. For me specifically, I am implmenting MapLibre for visualizing scans of paintings as single maps, and users have requested the ability to "fully zoom out" where they expect to see the entire painting.

**Comment by larsmaxfield** [comment]
Disabling the bounds error in the constructor for `CanonicalTileID` (`tile_id.ts`) allows 3D to work with under-zooming:

![under-zooming-3d](https://github.com/user-attachments/assets/1e81ddab-609f-432c-af8e-a4ddef7bcca5)

**Comment by HarelM** [comment]
I would say that for paintings you can have them shown in a larger zoom, can't you? i.e. show full painting in zoom 3, or even 7 would solve this, wouldn't it?

**Comment by larsmaxfield** [comment]
@HarelM ah, do you mean by padding the original images with a wide border so that the generated tilesets have plenty of empty space around the actual scan?

**Comment by sbachinin** [comment]
Not sure if it's the same as what Harel means but -
if you don't stretch the paintings to the whole globe but render them somewhat smaller in the center of the globe, it must be better in every way. E.g., the paintings' aspect ratio can be any.
I honestly don't know how to implement this but it must be doable

**Comment by larsmaxfield** [comment]
I use `pyvips` `dzsave` which tiles the painting by scaling the original image by half until it fits within a single tile, which becomes the lowest zoom 0/0/0.png image. The z-depth then depends on how many times the image was halved, or rather how large the original image was.

I hadn't considered rendering it smaller than the map. Perhaps because of how `pyvips` does the tiling. For me it makes more sense that the painting always fills the 0/0/0.png tile. I don't need to keep track of minimum zoom or bounds.

I have considered centering the painting, but for coordinate transformation that adds one more step (an XY offset). Assuming 0,0 is 0,0 in the painting is just easier.

**Comment by HarelM** [comment]
You can use image source instead of tiling I believe.

**Comment by larsmaxfield** [comment]
The scans are of gigapixel magnitude — 100k × 100k px — so we rely on tiling.

**Comment by sbachinin** [comment]
In my opinion, the idea of "underzooming" (viewing the whole globe on a single-globe map) totally makes sense.
It feels better to be able to zoom out enough to view the whole thing.
So if this can be implemented without much complexity, why not?
I also think that this can be the default behavior of a single-globe map, that is, I wouldn't introduce this new option because it looks a little odd.
I don't think that anyone is going to suffer from having blank margins around the globe.

As for the complexity, I doubt that you can make it work well with little code.
E.g., I think it should behave like this example from OpenSeadragon. That is, panning should still have some constraints (it must be impossible to outpan the map from the viewport completely). This is perhaps not a trivial task.
Also there is a lot of stuff that (potentially) can be broken by this change, like markers and popups.

**Comment by larsmaxfield** [comment]
Good points. I agree that this should become the new default behavior for single-globe maps.

I'll give it a try and update here when I've got something working.

**Comment by larsmaxfield** [comment]
I've proposed a solution in PR #4612 (quoted below), but there are two points I need help on:

- I do not know if the out-of-maxbounds behavior should be considered a breaking change. By setting renderWorldCopies=true (or simply not setting it at all) then maxBounds will behave as before. Also, if maxBounds is being set, I can't foresee a case where one would care to set renderWorldCopies false.

- I do not know what new tests (if any) are appropriate for this.

> Here is a demo where you can toggle `renderWorldCopies`:
> 
> ![unbounded](https://private-user-images.githubusercontent.com/83759569/361728300-a3cf3280-2051-4e28-933f-ddad87e6aabb.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ3NTI4OTMsIm5iZiI6MTcyNDc1MjU5MywicGF0aCI6Ii84Mzc1OTU2OS8zNjE3MjgzMDAtYTNjZjMyODAtMjA1MS00ZTI4LTkzM2YtZGRhZDg3ZTZhYWJiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODI3VDA5NTYzM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWU1NTJiYmExMDM1YzZmMDE1NGZjNDQyM2I4ZmUzMTFhMzdhODI5YjFhNjQ5NTZkNzk0NzdhZjY0OTAxMDNmMTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.NRuHu6smW5x9778_FGyX1NepiVoj0yjl9eUHgO4gFYo) https://larsmaxfield.github.io/swings/maplibre-underzoom/render-world-copies.html
> 
> The solution is a simple change to `getConstrained()` in `src/geo/transform.ts` where when `renderWorldCopies=false` the viewport (screen) is assumed to be a square with side length equal to the smallest screen dimension reduced by a `boundsRatio` factor of 50%. This allows the map to be panned and zoomed out 50% beyond the bounds along the narrowest screen dimension.
> 
> This change does not override `minZoom` — it always obeys that value.
> 
> Since users can pan and zoom 50% beyond the bounds, it means that a map with a specified `maxBounds` (and `renderWorldCopies=false`) will allow users to pan and zoom beyond those bounds. Here are two demos of that:
> 
> ![tall bounds](https://private-user-images.githubusercontent.com/83759569/361729269-0ff4566c-69d5-484d-b4cc-ff6a3724c0fe.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ3NTI4OTMsIm5iZiI6MTcyNDc1MjU5MywicGF0aCI6Ii84Mzc1OTU2OS8zNjE3MjkyNjktMGZmNDU2NmMtNjlkNS00ODRkLWI0Y2MtZmY2YTM3MjRjMGZlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODI3VDA5NTYzM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTUzMWQ0OGQ4ZmU2MTJjYjg0MDZlNDczMGY2YTk2NTY3YTZmYWQ4YTNiZDRiY2ZlMDVlMThkNDA2YTkzMDVlMDQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.6WVkKfABpa4TWChSWp-5G163NpAU3TndBAzc3nTNvMs) https://larsmaxfield.github.io/swings/maplibre-underzoom/tall-bounds.html
> 
> ![wide bounds](https://private-user-images.githubusercontent.com/83759569/361729638-00e10a07-776c-4298-b278-c13bfa77a2ed.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ3NTI4OTMsIm5iZiI6MTcyNDc1MjU5MywicGF0aCI6Ii84Mzc1OTU2OS8zNjE3Mjk2MzgtMDBlMTBhMDctNzc2Yy00Mjk4LWIyNzgtYzEzYmZhNzdhMmVkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODI3VDA5NTYzM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg1MzhlNWJmOWZkZDVjZThkZjQyY2ExNTU5ZWQ0ZGY5MDJkYThjZGQzYWRiMGM2MDI1NDMzZTliMzhmMDliNDgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.rB4wgnFSa2j0Zp8LFd3oMfEHhDhP3Mb9mE3CJt-Bw5w) https://larsmaxfield.github.io/swings/maplibre-underzoom/wide-bounds.html

**Comment by larsmaxfield** [comment]
Here is a demo of my new design that allows you to underzoom and overpan a single-copy world:

<sup>https://larsmaxfield.github.io/swings/maplibre-underzoom</sup>
[![Underzoom demo.](https://github.com/user-attachments/assets/0548100c-8313-432f-a648-a2b6a0c1580e)](
https://larsmaxfield.github.io/swings/maplibre-underzoom)

There are three parameters which I currently name as follows:

- `allowUnderzoom`: Boolean that enables/disables underzooming and overpanning.

- `underzoom`: Number 0–100 that defines how far you can underzoom the map bounds as a percent of the map's size relative to the viewport. If 100, you can underzoom the map until it is 100% the size of the viewport. If 50, the map can be underzoomed until it is 50% the size of the viewport.

- `overpan`: Number 0-50 that defines how far you can overpan the latitude-longitude bounds of the map as a percent of the viewport height and width. If 50, you can overpan the map until the latitude-longitude bounds reach the viewport center (half the height and width). If 25, you can overpan the map until the bounds reach 25% the viewport height and width. If 0, you cannot overpan the bounds.

The `minZoom` is always respected, meaning `underzoom = 0` will allow you to underzoom until `minZoom` is reached.

If you underzoom the map such that the bounds must exceed the limit of `overpan`, the map will allow that. In other words, `overpan` is overridden by `underzoom` when necessary. (Try setting `overpan = 0` and see that zooming out effectively centers the underzoomed map.)

The `maxBounds` is also respected insofar that underzoom and overpan will properly apply to those bounds:

- Wide area:
![Wide-bounded map with underzoom.](https://github.com/user-attachments/assets/4b3f193d-2f1f-46f0-b061-ac04ece1d623)

- Tall area:
![Tall-bounded map with underzoom.](https://github.com/user-attachments/assets/e78f1550-a4ff-44ea-9114-f7c1d9b7dacc)

**Comment by larsmaxfield** [comment]
> Why not use minZoom and bounds to define this behavior instead of introducing new configurations?

I considered that, but:

`minZoom` is based on absolute tile size, not the relative viewport size. This means that if I wanted to have a consistent underzooming size (say, allow a user to underzoom the map to 50% the viewport size), I would need to monitor the viewport size and update `minZoom` every time it changes.

`bounds` cannot exceed the built-in lnglat maximums. This makes it impossible to underzoom a map on a tall viewport.

**Comment by HarelM** [comment]
I tend to think that the definitions added here are not in line with the previous configurations. 
I would consider thinking which APIs this features needs and write it down as a plugin...

**Comment by larsmaxfield** [comment]
> I tend to think that the definitions added here are not in line with the previous configurations. 
> I would consider thinking which APIs this features needs and write it down as a plugin...

Could you explain how they're not in line? Or point to a good example I could learn from?

If it helps, the approach I took here is to default the map to current MapLibre behavior — no underzoom and no overpan unless both `renderWorldCopies=false` and `allowUnderzoom=true`. Also, the only function my implemention alters is `Tranform.getConstrained()`.

Good advice to see how a plugin could solve this. I'm still getting familiar with everything that can be done with MapLibre and hadn't considered that.

Nevertheless I agree with @sbachinin that underzooming could be the default single-world behavior.

**Comment by HarelM** [comment]
Most options are not in the range of 0 to 1, see here:
https://maplibre.org/maplibre-gl-js/docs/API/type-aliases/MapOptions/
I don't have a better suggestion though, unfortunately...

**Comment by larsmaxfield** [comment]
I updated the number options to be percentages with `underzoom` in the range 0–100 and `overpan` in the range 0–50 (see edited explanation above). I also made a draft PR to check the impact of the design.

Still looking into the plugin route.

**Comment by a-milenin** [comment]
Is there any news? This feature is really important.

**Comment by larsmaxfield** [comment]
@a-milenin indeed I just now ported my v4 PR to v5 #5995.

[Updated example](https://larsmaxfield.github.io/swings/maplibre-underzoom).

Still need to make tests.

Also still need approval that the options and design are OK.

**Comment by HarelM** [comment]
I've re-read the entire thread again and looked at the API added as part of the PR and I still find it a bit odd - The API and style spec is based on zoom, pixels, latitude and longitude. Introducing something that is not taking zoom etc into consideration is... well... odd, to me...
While the arguments here https://github.com/maplibre/maplibre-gl-js/issues/4510#issuecomment-2364070439 are relevant I haven't fully got on boarded, and the original question around which public API can be added in order to support this behavior maybe outside maplibre still remains - e.g. provide a constraint method similar to the cameraUpdateTransfrom to facilitate for this if other API can't facilitate for this (this is not a deeply though of suggestion, just from the top of my head).

**Comment by larsmaxfield** [comment]
I agree that this adds options that are quite foreign to what exists.

Instead of the terms underzoom and overpan, we could rename this feature and its options by what it actually does: Allow a user to change how the bounds are respected by the viewport.

The viewport currently respects the bounds by letting users pan and zoom until the viewport's edges reach the bounds. That means the viewport will not show what is beyond the bounds. This makes sense if we interpret the bounds as a limit on what users can _see_.

Alternatively we could interpret the bounds as a limit on where users can _move_ with respect to the viewport center.

(I would argue this alternative interpretation makes methods like `setCenter()` and `setZoom()` more predictable. If we call `setCenter()` to a valid location within the bounds but where the viewport would violate the current edge-to-bounds behavior, the map won't actually move to that exact location, but to somewhere nearby. The same happens with `setZoom()`; we might call `setZoom(4)`, but because of how the bounds and viewport are sized, it might only zoom out to 5 or some oddly non-integer value like 4.3562. This also means `minZoom` is sometimes not relevant, because we might not even be capable of zooming out that far. Even the tests reflect this unpredictability. All of that is solved by this feature.)

Back to rebranding: "Overpan" is really just letting users pan until the bounds touch the center of the viewport. "Underzoom" is really just letting users zoom out until the bounds are a certain size relative to the viewport. minZoom is always respected.

Thus we could rename the option(s) to reflect something like "viewport bounds behavior", where the default would be "viewport edge" and optionally changed to involve "viewport center". In layman's terms, that option says "let me move the viewport until the bounds reach the center, not just the edge."

For the granular control of overpan and underzoom I propose in the PR, maybe there are better terms for that. Though what users experience is an underzooming and overpanning of the bounds. I'm not very creative here.

To your suggestion about the public API and adding a constraint method, I don't have the experience or the immediate time to look into that. From what I investigated earlier in possible solutions, it all seems to come down to what `getConstrained()` does. Therefore any change would need to modify that function and/or add parameters inside for us to achieve this feature. And that's why my PR does just that. But maybe there's a different way to change `getConstrained()`.

Someone with time and a more educated eye would be very welcome to help here and double check all this.

**Comment by dhruvatara** [comment]
@larsmaxfield I found the issue that constrained the panning on a single-world map: https://github.com/maplibre/maplibre-gl-js/pull/3738

**Comment by larsmaxfield** [comment]
@dhruvatara that indeed relates to constraining horizontal panning, but this feature request includes vertical panning as well.

**Comment by a-milenin** [comment]
Is there any solution?

**Comment by larsmaxfield** [comment]
@a-milenin not yet.

@HarelM what do you think of this naming?

```ts
viewportBoundsContraint: 'confine' | 'extend',
extendScaleLimit?: number,
extendPanLimit?: number
```

I think the terms are clearer now:
- Viewport movement is **constrained** by the bounds
- By default, we **confine** the viewport within the bounds. We cannot zoom or pan to show areas outside the bounds.
- Optionally, we can **extend** the viewport beyond the bounds to certain relative limits. We can:
  - Zoom out until the bounds are a relative size to the viewport
  - Pan until the bounds are a relative distance to the viewport center

For example:

```js
const map = new maplibregl.Map({
    renderWorldCopies: false,  // Extend only if single non-globe world
    ...,
    viewportBoundsContraint: 'extend',
    extendScaleLimit: 50,  // You can zoom out until the bounds are 50% the viewport size
    extendPanLimit: 25  // You can pan until the bounds are 25% the distance to viewport center
    });

**Comment by HarelM** [comment]
I've seen recently some other requests around constraining the map, so I think adding more and more API for this would be the wrong solution. A general hook to override the map's constrains seems like the right approach to generalize the solution for this problem and others, I believe.

**Comment by larsmaxfield** [comment]
Ok, I'll take a look at that.

**Comment by HarelM** [comment]
I believe this is now possible using the new `transformConstrain` API, right? Can I close this issue?

**Comment by jo-chemla** [comment]
Thanks for this PR for @larsmaxfield, looking great from a user and dev perspective!

Your [demo](https://larsmaxfield.com/swings/maplibre-transform-constrain/) also do look great, with the parametrized `underzoomTransformConstrain` - via `scale` and `pan` underzooming ratios params. Would it be useful to add this underzoom constraint as an example rather or in addition to the identity `transformConstrain`? 



<details><summary>underzoomTransformConstrain from lars demo</summary>
<p>

```javascript

    let _extend = true;
    let _extendScalePercent = 70; 
    let _extendPanPercent = 0;

    function underzoomTransformConstrain(lngLat, zoom) {
        zoom = clamp(+zoom, this._minZoom, this._maxZoom);
        const result = {
            center: new maplibregl.LngLat(lngLat.lng, lngLat.lat),
            zoom
        };

        let lngRange = this._lngRange;

        if (!this._renderWorldCopies && lngRange === null) {
            const almost180 = 180 - 1e-10;
            lngRange = [-almost180, almost180];
        }

        const worldSize = this.tileSize * zoomScale(result.zoom); // A world size for the requested zoom level, not the current world size
        let minY = 0;
        let maxY = worldSize;
        let minX = 0;
        let maxX = worldSize;
        let scaleY = 0;
        let scaleX = 0;
        const {x: screenWidth, y: screenHeight} = this.size;

        const underzoom =  // 0-1 (percent as normalized factor of viewport minimum dimension)
            _extend ?
                clamp(_extendScalePercent, 0, 100) / 100 :
                1.0;

        if (this._latRange) {
            const latRange = this._latRange;
            minY = mercatorYfromLat(latRange[1]) * worldSize;
            maxY = mercatorYfromLat(latRange[0]) * worldSize;
            const shouldZoomIn = maxY - minY < (underzoom * screenHeight);
            if (shouldZoomIn) scaleY = underzoom * screenHeight / (maxY - minY);
        }

        if (lngRange) {
            minX = wrap(
                mercatorXfromLng(lngRange[0]) * worldSize,
                0,
                worldSize
            );
            maxX = wrap(
                mercatorXfromLng(lngRange[1]) * worldSize,
                0,
                worldSize
            );

            if (maxX < minX) maxX += worldSize;

            const shouldZoomIn = maxX - minX < (underzoom * screenWidth);
            if (shouldZoomIn) scaleX = underzoom * screenWidth / (maxX - minX);
        }

        const {x: originalX, y: originalY} = projectToWorldCoordinates(worldSize, lngLat);
        let modifiedX, modifiedY;

        const scale =
            _extend ?
                Math.min(scaleX || 0, scaleY || 0) :
                Math.max(scaleX || 0, scaleY || 0);

        if (scale) {
            // zoom in to exclude all beyond the given lng/lat ranges
            const newPoint = new maplibregl.Point(
                scaleX ? (maxX + minX) / 2 : originalX,
                scaleY ? (maxY + minY) / 2 : originalY);
            if (this._renderWorldCopies) result.center = unprojectFromWorldCoordinates(worldSize, newPoint).wrap();
            result.zoom += scaleZoom(scale);
            return result;
        }

        // Panning up and down in latitude is externally limited by project() with MAX_VALID_LATITUDE.
        // This limit prevents panning the top and bottom bounds farther than the center of the viewport.
        // Due to the complexity and consequence of altering project() or MAX_VALID_LATITUDE, we'll simply limit
        // the overpan to 50% the bounds to match that external limit.
        let lngOverpan = 0.0;
        let latOverpan = 0.0;
        if (_extend) {
            const overpan = clamp(_extendPanPercent, 0, 100) / 100;  // 0-1 (percent as a normalized factor from viewport edge to center)
            const latUnderzoomMinimumPan = 1.0 - ((maxY - minY) / screenHeight);
            const lngUnderzoomMinimumPan = 1.0 - ((maxX - minX) / screenWidth);
            lngOverpan = Math.max(lngUnderzoomMinimumPan, overpan);
            latOverpan = Math.max(latUnderzoomMinimumPan, overpan);
        }
        const lngPanScale = 1.0 - lngOverpan;
        const latPanScale = 1.0 - latOverpan;

        if (this._latRange) {
            const h2 = latPanScale * screenHeight / 2;
            if (originalY - h2 < minY) modifiedY = minY + h2;
            if (originalY + h2 > maxY) modifiedY = maxY - h2;
        }

        if (lngRange) {
            const centerX = (minX + maxX) / 2;
            let wrappedX = originalX;
            if (this._renderWorldCopies) {
                wrappedX = wrap(originalX, centerX - worldSize / 2, centerX + worldSize / 2);
            }
            const w2 = lngPanScale * screenWidth / 2;

            if (wrappedX - w2 < minX) modifiedX = minX + w2;
            if (wrappedX + w2 > maxX) modifiedX = maxX - w2;
        }

        // pan the map if the screen goes off the range
        if (modifiedX !== undefined || modifiedY !== undefined) {
            const newPoint = new maplibregl.Point(modifiedX ?? originalX, modifiedY ?? originalY);
            result.center = unprojectFromWorldCoordinates(worldSize, newPoint).wrap();
            if (this._renderWorldCopies) result.center = result.center.wrap();
        }

        return result;
    }

    function identityTransformConstrain(lngLat, zoom) {
        return {center: lngLat, zoom: zoom ?? 0};
    };

    const map = new maplibregl.Map({
        container: 'map',
        renderWorldCopies: false,
        transformConstrain: underzoomTransformConstrain,
        style: {
          //...
        },
    });
``` 

</p>
</details>

**Comment by HarelM** [comment]
I think a plugin would be a better fit to this logic, this is also the reason for making this API public.

**Comment by larsmaxfield** [comment]
Precisely.

@jo-chemla the [maplibre-xy](https://github.com/larsmaxfield/maplibre-xy) plugin I'm working on includes an updated underzoom transformConstrain. I plan to make it an NPM package, but you can already copy the `/src` to use it.

**Comment by larsmaxfield** [comment]
You can now install [maplibre-xy](https://github.com/larsmaxfield/maplibre-xy) from NPM or directly import its modules from unpkg.

Assign the underzooming transformConstrain with an `Underzoom` instance:

```js
import { Underzoom } from 'maplibre-xy';
// or
import { Underzoom } from 'https://unpkg.com/maplibre-xy';

const myUnderzoom = new Underzoom(maplibregl);  // Use default options

const map = new maplibregl.Map({
    transformConstrain: myUnderzoom.transformConstrain,
    ...
```
